### PR TITLE
Prepare 2.4.0 release

### PR DESCRIPTION
**Pull request by nicosomb**
🚀

### CODE REVIEW

**Comment by j0k3r** [comment]
Switched to WIP until be merge all PR related to 2.4.0

**Comment by di72nn** [comment]
I think 3158 should be moved/duplicated to the API section (since it adds new parameters to an endpoint).
3630, 3271 (add properties) and 4028 (adds an enpoint) should also be present in the API section.

4058 is more of UI/UX than API.

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,5 +1,87 @@
...
+- Cascade delete on oauth2 table when deleting a user [#3956](https://github.com/wallabag/wallabag/pull/3956)
+- Preview picture: use the 1st pic retrieved if no og:image set [#3965](https://github.com/wallabag/wallabag/pull/3965)
+- Force PHP version in Dockerfile [#3977](https://github.com/wallabag/wallabag/pull/3977)
+- Add Sentry support [#3990](https://github.com/wallabag/wallabag/pull/3990)
+- Fix slow/unindexed queries [#4081](https://github.com/wallabag/wallabag/pull/4081)
+- Rename internal settings table [#4084](https://github.com/wallabag/wallabag/pull/4084)
+- Add environment variables to control scripts/dev.sh behavior [#4152](https://github.com/wallabag/wallabag/pull/4152)
+- Increase php memory limit PHP Docker [#4163](https://github.com/wallabag/wallabag/pull/4163)
+- Ensure ENV is well defined when using make [#4249](https://github.com/wallabag/wallabag/pull/4249)
+- Ensure ENV is well defined when using make [#4249](https://github.com/wallabag/wallabag/pull/4249)
```
Comment (tcitworld) [inline CHANGELOG.md:66]: Duplicate

**Comment by j0k3r** [comment]
> I think 3158 should be moved/duplicated to the API section (since it adds new parameters to an endpoint).
> 3630, 3271 (add properties) and 4028 (adds an enpoint) should also be present in the API section.
> 
> 4058 is more of UI/UX than API.

Thanks for suggestion, I've moved what I think it should be moved. Also I don't want to duplicate information between different section.

**Comment by j0k3r** [comment]
👋 Draft for 2.4.0 release note 👇
Related: https://github.com/wallabag/wallabag/projects/4#card-40158216

---
🎉 **The long awaited 2.4.0 is finally here** 🌟 
There are a ton a new features (you can check the full changelog if you want to see them all). We'll only highlight main ones here.

# New features

- Allow to rename tags from the web interface [#3574](https://github.com/wallabag/wallabag/pull/3574)
- Possibility to change locale from login/register pages [#3216](https://github.com/wallabag/wallabag/pull/3216)
- Ability to import/export tagging rules [#4028](https://github.com/wallabag/wallabag/pull/4028)
- Enable OTP 2FA via third apps (Google Authenticator, Authy or FreeOTP) [#3798](https://github.com/wallabag/wallabag/pull/3798)
- Add random feature [#3526](https://github.com/wallabag/wallabag/pull/3526)
- Ability to revoke feed token [#3994](https://github.com/wallabag/wallabag/pull/3994)
- Move Ignore Origin rules to database [#4026](https://github.com/wallabag/wallabag/pull/4026)
- Ability to manually define the reading speed [#4053](https://github.com/wallabag/wallabag/pull/4053)
- MathJax support [#4128](https://github.com/wallabag/wallabag/pull/4128)
- Allow custom styles system wide [#4151](https://github.com/wallabag/wallabag/pull/4151)
- Support elCurator import [#4175](https://github.com/wallabag/wallabag/pull/4175)
- Added mass actions for Material design in list view [#4326](https://github.com/wallabag/wallabag/pull/4326)

# Technical changes

## Support for PHP 7.4

The 2.3.8 lacked support for PHP 7.4 but, hey, that wallabag 2.3.8 was released in May 2019 and PHP 7.4 first alpha goes out in June 2019.
Maybe we should have release 2.3.9 with PHP 7.4 support but there were too much stuff in the master and lack of motivation to only keep commits which enabled support.
I hope we'll be more reactive for PHP 8.0 :)

## Dropped support for PHP < 7.2

_PRs #3758 #4441_

We removed support for PHP < 7.2 because lower version aren't supported by PHP itself.
First we remove support for PHP < 7.1 back in Novembre 2018 https://github.com/wallabag/wallabag/pull/3758 and then support for PHP < 7.2 in June 2020.

Some people will arg that we should keep the lowest version possible because some folks are maintaining outdated version of PHP but we need to move on.
Also, security support for PHP 7.2 will end in [November 2020](https://www.php.net/supported-versions.php) but we think it's too fast for most people. Also Symfony still required PHP >=7.2.5 for the current master (the upcoming 5.2.0).

## Translations in Weblate

_PRs #4447 #4359 #4442_

The whole application is now connected to the Weblate platform which allow anyone to contribute to the translation without having to use Git at all.
There are [multiple sources in Weblate for the wallabag project](https://hosted.weblate.org/projects/wallabag/) because the iOS & Android app are also using it.

The application is now fully translated in 10 languages! :fr: :us: :cn: :kr: :jp: :netherlands: :ru: :tr: :de: :es:
It's amazing.
✏️ Feel free to improve translations 

## Hashed URL for fastest check

_PR #3158_

People using the API to save contents often have to check if the URL already exists in the database. Using the plain URL to check wasn't fast enough.
That why there is now a hash of the URL for each entry. Don't forget to run the command `php bin/console --env=prod wallabag:generate-hashed-urls` to generate hash for URLs saved before your 2.4.0 upgrade.

URL are hashed using `sha1`.

## Add basic search endpoint

_PR #3627_

The API now expose a _basic_ search endpoint. We declare it as basic because it's a simple search in the database (like the one in the web interface). No robust search engine were involved in that change.

# How to upgrade

- perform a backup of the database (just in case)
- run `make update` from your wallabag directoy
- run `php bin/console --env=prod wallabag:generate-hashed-urls` which will generates a hash of URL for all saved entries (improve API search+++)

You might be interesting in these information too:
- you can safely remove `database_driver_class` from the `app/config/parameters.yml` file
- install `imagick` PHP extension if you want GIF animation to be kept when saving images locally
- you can now use [Sentry](https://sentry.io/) to log errors from wallabag, just define `sentry_dsn` in `app/config/parameters.yml`

**Comment by di72nn** [comment]
Not to be nitpicky, but here are some corrections for you to consider:

> There are a ton a new features

There's a ton of new features

> via third apps

via third-party apps

> there were too much stuff

there was too much stuff

> Feel free to improve translations

The lack of period really bothers me.

> Hashed URL for fastest check

Hashed URLs for faster check

> That why

That's why

> URL are hashed

URLs are hashed

> might be interesting in these information

might be interested in this information

I would also appreciate list items to end in periods or commas.

**Comment by j0k3r** [comment]
@di72nn no problem, changes applied!

**Comment by tcitworld** [comment]
> generate hash for URLs saved before your 2.4.0 upgrade

It should say "after", since that command is available in 2.4.0.

**Comment by j0k3r** [comment]
@tcitworld is it more clear?
image

**Comment by 00dani** [comment]
While hashing the URLs of entries *does* make checking which URLs are saved faster, I think that's not the most important benefit?

As I discussed on wallabag/wallabagger#204, there's a significant *privacy* benefit to using hashed URLs: it becomes possible to discover whether a given URL is saved in Wallabag without actually sending that URL to your Wallabag server. This is especially important if you're using a shared server like wallabag.it, which you mightn't want to trust with a complete log of every URL you ever visit.

I think mentioning this privacy benefit in the changelog would be a good idea, especially since it'll encourage API consumers to use the new feature. 🐱

**Comment by j0k3r** [comment]
I've updated the release note to add message about baggy, because it'll be deprecated and then remove in 2.5.0.

**Review by j0k3r — APPROVED** [review]

**Review by tcitworld — APPROVED** [review]
It's happening 🎉

**Review by Kdecherf — APPROVED** [review]

**Pull request merged**